File: legacy/src/scenes/dungeon/debug-spawn.ts
/**
 * DEBUG SPAWN LAYOUT — where a scripted spawn actually puts its monsters.
 *
 * The floor already had a spawn debugger: the ` panel's enemy chips, one click
 * one monster, dropped next to the knight. That is fine for eyeballing a sprite
 * and useless for testing anything else, because almost every question worth
 * asking needs a CROWD in a KNOWN SHAPE:
 *
 *   · does Blade Storm's ring actually bite everything inside its radius?
 *   · does Time Crawl visibly smear the horde? (needs a horde)
 *   · does the Arcane Pulse wave front strike each foe as it crosses it, in
 *     distance order? (needs foes at known distances)
 *   · does an AoE stop at its edge? (needs one just inside and one just outside)
 *
 * So the layout is the part worth extracting and testing: this module is pure
 * (no three, no DOM, no game state) and answers "given a centre and a spec,
 * which WALKABLE tiles do the monsters go on". core.ts owns actually building
 * them; `__dungeonSpawn` is the harness-facing hook.
 *
 * A ring is snapped tile-by-tile rather than placed at exact polar coordinates:
 * a monster inside a wall is worse than a monster a tile off its ideal angle,
 * and a maze rarely has clean floor at every bearing.
 */
import { type Grid, type TilePos, isWalkable, worldToTile, tileCenter } from "./maze/generator";

/** How a scripted spawn arranges its monsters. */
export interface SpawnLayout {
  /** How many to place. */
  count: number;
  /**
   * Distance from the centre, in tiles. 0 (or absent) = pack them as close to
   * the centre as open floor allows — the old chip behaviour, but for N.
   * A ring is what you want for AoE work: every monster at a known range.
   */
  ring?: number;
  /** Rotate the ring, so successive calls don't stack on the same bearings. */
  phase?: number;
}

/** The ideal (unsnapped) offsets for a layout, in tiles. Exported for tests and
 *  because a caller may want to know how far it *asked* for. */
export function layoutOffsets(layout: SpawnLayout): Array<{ di: number; dj: number }> {
  const n = Math.max(0, Math.floor(layout.count));
  const r = layout.ring ?? 0;
  const phase = layout.phase ?? 0;
  const out: Array<{ di: number; dj: number }> = [];
  for (let k = 0; k < n; k++) {
    if (r <= 0) {
      out.push({ di: 0, dj: 0 }); // all at the centre; the snap fans them out
      continue;
    }
    const a = phase + (k / n) * Math.PI * 2;
    out.push({ di: Math.cos(a) * r, dj: Math.sin(a) * r });
  }
  return out;
}

/**
 * The nearest walkable tile to (ti, tj) that nothing has claimed yet, searched
 * outward in square rings. Returns null only if the whole neighbourhood within
 * `maxR` is wall or taken — i.e. there is genuinely nowhere to stand.
 *
 * `taken` is threaded through the whole layout so two monsters never land on
 * one tile, which is what makes a "ring of 8" actually read as eight.
 */
export function freeTileNear(g: Grid, ti: number, tj: number, taken: Set<number>, maxR = 6): TilePos | null {
  for (let r = 0; r <= maxR; r++) {
    for (let dj = -r; dj <= r; dj++) {
      for (let di = -r; di <= r; di++) {
        if (r > 0 && Math.max(Math.abs(di), Math.abs(dj)) !== r) continue; // ring shell only
        const i = ti + di;
        const j = tj + dj;
        const key = j * g.w + i;
        if (taken.has(key) || !isWalkable(g, i, j)) continue;
        taken.add(key);
        return { i, j };
      }
    }
  }
  return null;
}

/** One resolved spawn slot: the tile chosen, and the world centre of it. */
export interface SpawnPoint {
  i: number;
  j: number;
  x: number;
  z: number;
}

/**
 * How much a candidate tile may miss the requested radius before we stop
 * considering it at all. Generous, because a maze corridor genuinely may not
 * have floor at the asked-for range in most directions.
 */
const RING_SEARCH_SLACK = 4;
/**
 * Cost weights when picking a tile for a ring slot. RADIUS DOMINATES BEARING,
 * and that ordering is the whole point of this function.
 *
 * The first version snapped each slot outward from its ideal POINT, which reads
 * fine in an open room and collapses completely in a corridor: every off-axis
 * bearing is wall, so each slot walks back to the nearest floor — which is the
 * lane the knight is standing in. Asking for "8 at radius 3" returned monsters
 * at 0.0, 1.0, 1.41… i.e. standing on him. Since the reason to ask for a ring
 * is to put a horde at a KNOWN RANGE (AoE reach, wave-front ordering), losing
 * the radius loses the feature; losing some angular spread does not.
 */
const RING_W_RADIUS = 3;
const RING_W_BEARING = 1;

/** Signed shortest angular distance between two bearings, in radians. */
function angleDelta(a: number, b: number): number {
  let d = (a - b) % (Math.PI * 2);
  if (d > Math.PI) d -= Math.PI * 2;
  if (d < -Math.PI) d += Math.PI * 2;
  return Math.abs(d);
}

/** Every walkable tile in the annulus we're willing to consider for a ring. */
function ringCandidates(g: Grid, ci: number, cj: number, r: number): Array<{ i: number; j: number; d: number; a: number }> {
  const out: Array<{ i: number; j: number; d: number; a: number }> = [];
  const reach = Math.ceil(r + RING_SEARCH_SLACK);
  for (let dj = -reach; dj <= reach; dj++) {
    for (let di = -reach; di <= reach; di++) {
      const d = Math.hypot(di, dj);
      if (Math.abs(d - r) > RING_SEARCH_SLACK) continue;
      const i = ci + di;
      const j = cj + dj;
      if (!isWalkable(g, i, j)) continue;
      out.push({ i, j, d, a: Math.atan2(dj, di) });
    }
  }
  return out;
}

/**
 * Resolve a layout around a WORLD position into distinct walkable tiles.
 *
 * Fewer points than `count` come back when the room genuinely cannot hold them
 * — the caller reports that rather than silently stacking monsters, so a test
 * that says "spawn 8" and gets 5 knows its floor was too tight instead of
 * quietly asserting against the wrong horde size.
 */
export function resolveSpawnPoints(g: Grid, cx: number, cz: number, layout: SpawnLayout): SpawnPoint[] {
  const centre = worldToTile(g, cx, cz);
  const taken = new Set<number>();
  const out: SpawnPoint[] = [];
  const r = layout.ring ?? 0;
  const push = (i: number, j: number): void => {
    const c = tileCenter(g, i, j);
    out.push({ i, j, x: c.x, z: c.z });
  };

  // No ring: pack them as tightly around the centre as open floor allows.
  if (r <= 0) {
    for (let k = 0; k < Math.max(0, Math.floor(layout.count)); k++) {
      const spot = freeTileNear(g, centre.i, centre.j, taken);
      if (spot) push(spot.i, spot.j);
    }
    return out;
  }

  // Ring: score every candidate per slot, radius first (see the weights above).
  const cand = ringCandidates(g, centre.i, centre.j, r);
  for (const off of layoutOffsets(layout)) {
    const want = Math.atan2(off.dj, off.di);
    let best: (typeof cand)[number] | null = null;
    let bestCost = Infinity;
    for (const c of cand) {
      const key = c.j * g.w + c.i;
      if (taken.has(key)) continue;
      const cost = Math.abs(c.d - r) * RING_W_RADIUS + angleDelta(c.a, want) * RING_W_BEARING;
      if (cost < bestCost) {
        bestCost = cost;
        best = c;
      }
    }
    if (!best) continue; // annulus exhausted — report short rather than stack
    taken.add(best.j * g.w + best.i);
    push(best.i, best.j);
  }
  return out;
}

File: crates/pk-core/src/spawn/debug_spawn.rs
//! Debug spawn layout — where a scripted spawn actually puts its monsters.
//!
//! Port of `legacy/src/game/pinball-knight/debug-spawn.ts` (217 lines).
//!
//! PORTS: `debug-spawn.ts`

use std::collections::HashSet;
use std::f64::consts::PI;

use crate::grid::{is_walkable, tile_center, world_to_tile, Grid};
use crate::maze::track_launch::TilePos;
use crate::monsters::types::EnemyKind;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct SpawnLayout {
    pub count: usize,
    pub ring: Option<f64>,
    pub phase: Option<f64>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DebugSpawnSpec {
    pub layout: SpawnLayout,
    pub kind: EnemyKind,
    pub hp: Option<i32>,
    pub aggro: bool,
    pub at: Option<(f64, f64)>,
    pub ztype: Option<String>,
}

impl Default for DebugSpawnSpec {
    fn default() -> Self {
        Self {
            layout: SpawnLayout {
                count: 1,
                ring: None,
                phase: None,
            },
            kind: EnemyKind::Zombie,
            hp: None,
            aggro: true,
            at: None,
            ztype: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct DebugSpawnResult {
    pub spawned: usize,
    pub requested: usize,
    pub kind: String,
    pub points: Vec<(f64, f64)>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Offset {
    pub di: f64,
    pub dj: f64,
}

pub fn layout_offsets(layout: &SpawnLayout) -> Vec<Offset> {
    let n = layout.count;
    let r = layout.ring.unwrap_or(0.0);
    let phase = layout.phase.unwrap_or(0.0);
    let mut out = Vec::with_capacity(n);

    for k in 0..n {
        if r <= 0.0 {
            out.push(Offset { di: 0.0, dj: 0.0 });
            continue;
        }
        let a = phase + ((k as f64) / (n as f64)) * PI * 2.0;
        out.push(Offset {
            di: a.cos() * r,
            dj: a.sin() * r,
        });
    }
    out
}

pub fn free_tile_near(
    g: &Grid,
    ti: i32,
    tj: i32,
    taken: &mut HashSet<usize>,
    max_r: i32,
) -> Option<TilePos> {
    for r in 0..=max_r {
        for dj in -r..=r {
            for di in -r..=r {
                if r > 0 && di.abs().max(dj.abs()) != r {
                    continue;
                }
                let i = ti + di;
                let j = tj + dj;
                let key = (j * g.w + i) as usize;
                if taken.contains(&key) || !is_walkable(g, i, j) {
                    continue;
                }
                taken.insert(key);
                return Some(TilePos { i, j });
            }
        }
    }
    None
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SpawnPoint {
    pub i: i32,
    pub j: i32,
    pub x: f64,
    pub z: f64,
}

const RING_SEARCH_SLACK: f64 = 4.0;
const RING_W_RADIUS: f64 = 3.0;
const RING_W_BEARING: f64 = 1.0;

fn angle_delta(a: f64, b: f64) -> f64 {
    let mut d = (a - b) % (PI * 2.0);
    if d > PI {
        d -= PI * 2.0;
    }
    if d < -PI {
        d += PI * 2.0;
    }
    d.abs()
}

#[derive(Debug, Clone, Copy)]
struct Candidate {
    i: i32,
    j: i32,
    d: f64,
    a: f64,
}

fn ring_candidates(g: &Grid, ci: i32, cj: i32, r: f64) -> Vec<Candidate> {
    let mut out = Vec::new();
    let reach = (r + RING_SEARCH_SLACK).ceil() as i32;
    for dj in -reach..=reach {
        for di in -reach..=reach {
            let d = ((di * di + dj * dj) as f64).sqrt();
            if (d - r).abs() > RING_SEARCH_SLACK {
                continue;
            }
            let i = ci + di;
            let j = cj + dj;
            if !is_walkable(g, i, j) {
                continue;
            }
            out.push(Candidate {
                i,
                j,
                d,
                a: (dj as f64).atan2(di as f64),
            });
        }
    }
    out
}

pub fn resolve_spawn_points(
    g: &Grid,
    cx: f64,
    cz: f64,
    layout: &SpawnLayout,
) -> Vec<SpawnPoint> {
    let centre = world_to_tile(g, cx, cz);
    let mut taken = HashSet::new();
    let mut out = Vec::new();
    let r = layout.ring.unwrap_or(0.0);

    let push = |out: &mut Vec<SpawnPoint>, i: i32, j: i32| {
        let c = tile_center(g, i, j);
        out.push(SpawnPoint {
            i,
            j,
            x: c.0,
            z: c.1,
        });
    };

    if r <= 0.0 {
        for _ in 0..layout.count {
            if let Some(spot) = free_tile_near(g, centre.0, centre.1, &mut taken, 6) {
                push(&mut out, spot.i, spot.j);
            }
        }
        return out;
    }

    let cand = ring_candidates(g, centre.0, centre.1, r);
    for off in layout_offsets(layout) {
        let want = off.dj.atan2(off.di);
        let mut best: Option<Candidate> = None;
        let mut best_cost = f64::INFINITY;
        for c in &cand {
            let key = (c.j * g.w + c.i) as usize;
            if taken.contains(&key) {
                continue;
            }
            let cost = (c.d - r).abs() * RING_W_RADIUS + angle_delta(c.a, want) * RING_W_BEARING;
            if cost < best_cost {
                best_cost = cost;
                best = Some(*c);
            }
        }
        if let Some(b) = best {
            taken.insert((b.j * g.w + b.i) as usize);
            push(&mut out, b.i, b.j);
        }
    }
    out
}
